woocommerce/woocommerce
# Block Hooks API: Add the block hooks algorithm to Woo templates

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

This adds the block hooks algorithm to WooCommerce block templates. Since core methods to build block templates are not public we implemented our own. This meant that we do not get enhancements that come from WordPress core, including the block hooks functionality.

This PR brings that functionality to WooCommerce blocks templates. So when developers add hooked blocks to their site they are inserted into Woo templates as well to maintain consistency, previously this did not happen.

Closes https://github.com/woocommerce/woocommerce/issues/45618



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Check out this PR & ensure you're running WP 6.5 latest using the Beta Testing plugin
2. Clear all template customizations on your site.
3. Add the first code block to your site. Check the Single Product template and check that the loginout button is inserted after the product details block in both the editor and on the frontend.
4. Save the template. Reload the editor and using the Code Editor check that the `core/loginout` block has been added to the `ignoredHookedBlocks` meta of the product details block. This confirms it works for block templates loaded from the file system.
5. Now add the second block of code to your site. Reload the Single Product template in both the site editor and the frontend and check that the breadcrumbs block has been automatically inserted into the template.
6. Make a change and save the template. Using the Code Editor check that the `woocommerce/breadcrumbs` block has been added to the `ignoredHookedBlocks` meta of the product details block. This confirms it works for customized templates loaded from the database.
7. Try this on a variety of different Woo templates and blocks to ensure it is working as expected.

```php
function hook_loginout_block( $hooked_blocks, $position, $anchor_block, $context ) {
	if ( $anchor_block === 'woocommerce/product-details' && $position === 'after' ) {
		$hooked_blocks[] = 'core/loginout';
	}

	return $hooked_blocks;
}

add_filter( 'hooked_block_types', 'hook_loginout_block', 10, 4 );
```

```php
function hook_breadcrumbs_block( $hooked_blocks, $position, $anchor_block, $context ) {
	if ( $anchor_block === 'woocommerce/product-details' && $position === 'after' ) {
		$hooked_blocks[] = 'woocommerce/breadcrumbs';
	}

	return $hooked_blocks;
}

add_filter( 'hooked_block_types', 'hook_breadcrumbs_block', 10, 4 );
```



### Changelog entry




<details>

#### Significance




#### Type




#### Message Adds block hooks algorithm to WooCommerce templates.

#### Comment 

</details>
